power_grid_model_c/power_grid_model/include/power_grid_model/component/sensor.hpp

```diff
@@ -17,12 +17,19 @@
 
 namespace power_grid_model {
 
+template <template <bool symmetric> typename SensorCalcParam>
+    requires sensor_calc_param_type<SensorCalcParam<true>> && sensor_calc_param_type<SensorCalcParam<false>>
```

Comment:
Do we really need CRTP here? Mixing CRTP and virtual functions is pretty confusing.

I see the reason you use CRTP is now we have different return type for `sym_calc_param` and `asym_calc_param`. I would propose:

1. remove these two pure virtual functions at `Sensor`.
2. define new pure virtual functions at `GenericVoltageSensor` and `GenericPowerSensor`. At that time the return type is known.

Reply:
that's not possible because we need `calc_param` to exist here, including return type. or we need to step away from the notion that all sensors have calculation params. `calc_param` cannot be virtual without a return type. any thoughts?

Reply:
Do we have to require `calc_param` to exist at `Sensor` level?

We only iterate of all `GenericVoltageSensor` or `GenericPowerSensor`. But never all `Sensor`.

Reply:
will remove

Reply:
i did the alternative implementation and it introduces a lot of additional rule-of-five logic. imho, it gets more ugly instead of less. also not sure why but running into some linking errors that seem hard to resolve. are you really strongly against this?

Reply:
instead of Sensor being abstract, I went with a solution where I made all constructors and the destructor `protected`, meaning it's a class that is not abstract but it is something not constructible. statically ensuring that it's not `copyable` and also not constructible/copyable from derived classes